File: scripts/ghidra/import_bank_ff_listing.py
# Import labels/comments from bank_FF.asm into current Ghidra program.
#@category Pacman
#@keybinding
#@menupath
#@toolbar

from ghidra.program.model.listing import CodeUnit
import re

LABEL_RE = re.compile(r'^\s*([A-Za-z_][A-Za-z0-9_]*)\s*:\s*$')
ADDR_RE = re.compile(r'0x([0-9A-Fa-f]{6})\s+00:([0-9A-Fa-f]{4}):')
COMMENT_RE = re.compile(r';\s*(.+)$')


def _load_lines(path):
    f = open(path, 'r')
    try:
        return f.readlines()
    finally:
        f.close()


def _is_code_symbol(name):
    prefixes = ('vec_', 'sub_', 'loc_', 'ofs_', 'off_')
    return name.startswith(prefixes)


def _maybe_create_function(name, addr):
    if not (name.startswith('vec_') or name.startswith('sub_')):
        return False
    if getFunctionAt(addr) is not None:
        return False

    # Best effort: disassemble first, then create function.
    try:
        disassemble(addr)
    except Exception:
        pass

    try:
        createFunction(addr, name)
        return True
    except Exception:
        return False


args = getScriptArgs()
listing_path = 'bank_FF.asm'
if len(args) >= 1 and args[0]:
    listing_path = args[0]

lines = _load_lines(listing_path)

pending_labels = []
labels_created = 0
functions_created = 0
comments_added = 0

for raw in lines:
    line = raw.rstrip('\n')

    lm = LABEL_RE.match(line)
    if lm:
        pending_labels.append(lm.group(1))
        continue

    am = ADDR_RE.search(line)
    if not am:
        continue

    cpu_addr = int(am.group(2), 16)
    addr = toAddr(cpu_addr)

    if pending_labels:
        for name in pending_labels:
            try:
                createLabel(addr, name, True)
                labels_created += 1
            except Exception:
                # Ignore duplicate / invalid names.
                pass

            if _is_code_symbol(name) and _maybe_create_function(name, addr):
                functions_created += 1
        pending_labels = []

    cm = COMMENT_RE.search(line)
    if cm:
        comment = cm.group(1).strip()
        if comment:
            cu = getCodeUnitAt(addr)
            if cu is not None:
                old = cu.getComment(CodeUnit.EOL_COMMENT)
                if old is None or len(old.strip()) == 0:
                    cu.setComment(CodeUnit.EOL_COMMENT, comment)
                    comments_added += 1

print('Listing imported: %s' % listing_path)
print('Labels created: %d' % labels_created)
print('Functions created: %d' % functions_created)
print('Comments added: %d' % comments_added)

File: scripts/ghidra/ImportBankFfListing.java
// Import labels/comments from bank_FF.asm into current Ghidra program.
//@category Pacman

import ghidra.app.script.GhidraScript;
import ghidra.program.model.symbol.SourceType;
import ghidra.program.model.listing.Function;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.Listing;

import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ImportBankFfListing extends GhidraScript {

    private static final Pattern LABEL_RE = Pattern.compile("^\\s*([A-Za-z_][A-Za-z0-9_]*)\\s*:\\s*$");
    private static final Pattern ADDR_RE = Pattern.compile("0x([0-9A-Fa-f]{6})\\s+00:([0-9A-Fa-f]{4}):");
    private static final Pattern COMMENT_RE = Pattern.compile(";\\s*(.+)$");

    private boolean isCodeSymbol(String name) {
        return name.startsWith("vec_") ||
               name.startsWith("sub_") ||
               name.startsWith("loc_") ||
               name.startsWith("bra_") ||
               name.startsWith("ofs_") ||
               name.startsWith("off_");
    }

    private void tryDisassemble(Address addr) {
        try {
            disassemble(addr);
        }
        catch (Exception ignored) {
            // Best effort.
        }
    }

    private boolean maybeCreateFunction(String name, Address addr) {
        if (!(name.startsWith("vec_") || name.startsWith("sub_"))) {
            return false;
        }

        Function existing = getFunctionAt(addr);
        if (existing != null) {
            if (!name.equals(existing.getName())) {
                try {
                    existing.setName(name, SourceType.USER_DEFINED);
                }
                catch (Exception ignored) {
                    // keep existing name
                }
            }
            return false;
        }

        Listing listing = currentProgram.getListing();
        Instruction ins = listing.getInstructionAt(addr);
        if (ins == null) {
            try {
                clearListing(addr, addr);
            }
            catch (Exception ignored) {
                // best effort
            }
        }
        tryDisassemble(addr);

        try {
            return createFunction(addr, name) != null;
        }
        catch (Exception ignored) {
            return false;
        }
    }

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        String listingPath = "bank_FF.asm";
        if (args.length >= 1 && args[0] != null && !args[0].isEmpty()) {
            listingPath = args[0];
        }

        List<String> pendingLabels = new ArrayList<>();
        int labelsCreated = 0;
        int functionsCreated = 0;
        int commentsAdded = 0;

        try (BufferedReader br = new BufferedReader(new FileReader(listingPath))) {
            String line;
            while ((line = br.readLine()) != null && !monitor.isCancelled()) {

                Matcher lm = LABEL_RE.matcher(line);
                if (lm.matches()) {
                    pendingLabels.add(lm.group(1));
                    continue;
                }

                Matcher am = ADDR_RE.matcher(line);
                if (!am.find()) {
                    continue;
                }

                int cpuAddr = Integer.parseInt(am.group(2), 16);
                Address addr = toAddr(cpuAddr);

                if (!pendingLabels.isEmpty()) {
                    for (String name : pendingLabels) {
                        try {
                            createLabel(addr, name, true, SourceType.USER_DEFINED);
                            labelsCreated++;
                        }
                        catch (Exception ignored) {
                            // Ignore duplicate / invalid label names.
                        }

                        if (isCodeSymbol(name) && maybeCreateFunction(name, addr)) {
                            functionsCreated++;
                        }
                        else if (isCodeSymbol(name)) {
                            tryDisassemble(addr);
                        }
                    }
                    pendingLabels.clear();
                }

                Matcher cm = COMMENT_RE.matcher(line);
                if (cm.find()) {
                    String comment = cm.group(1).trim();
                    if (!comment.isEmpty()) {
                        String old = getEOLComment(addr);
                        if (old == null || old.trim().isEmpty()) {
                            setEOLComment(addr, comment);
                            commentsAdded++;
                        }
                    }
                }
            }
        }

        println("Listing imported: " + listingPath);
        println("Labels created: " + labelsCreated);
        println("Functions created: " + functionsCreated);
        println("Comments added: " + commentsAdded);
    }
}
